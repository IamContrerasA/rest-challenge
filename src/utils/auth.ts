import jwt from 'jsonwebtoken'
import { PrismaClient, User } from '@prisma/client'
import { Request, Response, NextFunction } from 'express'
import bcrypt from 'bcrypt'

const prisma = new PrismaClient()
const JWT_SECRET = 'JWT_SECRET'

export const newToken = (user: User | null) => {
  return jwt.sign({ id: user?.id }, JWT_SECRET, {
    expiresIn: '100d',
  })
}

export const verifyToken = (token: string) =>
  new Promise((resolve, reject) => {
    jwt.verify(token, JWT_SECRET, (err, payload) => {
      if (err) return reject(err)
      resolve(payload)
    })
  })

export const signup = async (req: Request, res: Response): Promise<void> => {
  if (!req.body.email || !req.body.password) {
    res.status(400).send({ message: 'need email and password' })
  }

  try {
    let user: User
    bcrypt.hash(req.body.password, 8, async (err, hash) => {
      if (err) {
        throw new Error(`There is an error on hash password ${err}`)
      }
      user = await prisma.user.create({
        data: {
          email: req.body.email,
          password: hash,
        },
      })
    })

    const token = newToken(user!)
    res.status(201).send({ token })
  } catch (e) {
    return res.status(500).end()
  }
}

//checkpassword
function checkPassword(userPassord: string, password: string) {
  return new Promise((resolve, reject) => {
    bcrypt.compare(password, userPassord, (err, same) => {
      if (err) {
        console.log(err)
        return reject(err)
      }
      resolve(same)
    })
  })
}

export async function signin(req: Request, res: Response): Promise<void> {
  if (!req.body.email || !req.body.password) {
    res.status(400).send({ message: 'need email and password' })
  }

  const invalid = 'Invalid email and passoword combination'

  try {
    const user = await prisma.user.findUnique({
      where: { email: req.body.email },
    })

    if (!user) {
      res.status(401).send({ message: `${invalid}, no user found` })
    }

    const match = await checkPassword(user!.password, req.body.password)

    if (!match) {
      res.status(401).send({ message: `${invalid}, password dont match` })
    }

    const token = newToken(user)
    res.status(201).send({ token })
  } catch (e) {
    console.error(e)
    res.status(500).end()
  }
}

export const protect = async (
  req: Request,
  res: Response,
  next: NextFunction,
): Promise<void> => {
  const bearer = req.headers.authorization

  if (!bearer || !bearer.startsWith('Bearer ')) {
    return res.status(401).end()
  }

  const token = bearer.split('Bearer ')[1].trim()
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  let payload: any
  try {
    payload = await verifyToken(token)
  } catch (e) {
    return res.status(401).end()
  }

  const user = await prisma.user.findUnique({
    where: { id: payload.id },
    select: {
      id: true,
      firstname: true,
      surname: true,
      email: true,
      password: false,
    },
  })
  if (!user) {
    return res.status(401).end()
  }

  req.body.user = user
  next()
}
