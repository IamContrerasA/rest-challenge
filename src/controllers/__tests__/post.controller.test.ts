/* eslint-disable @typescript-eslint/no-explicit-any */
import { prisma, signup } from '../../utils/auth'
import {
  createPost,
  deletePost,
  findPost,
  findPosts,
  likedPost,
  publishPost,
  updatePost,
} from '../posts.controller'

const reqGlobal = {
  params: { id: 1 },
  body: {
    title: 'title of fake post',
    content: 'content of fake post',
    user: {
      id: 1,
    },
  },
}

beforeAll(async () => {
  const req: any = {
    body: {
      email: 'myemail14@email.com',
      password: 'password14',
    },
  }

  const res: any = {
    status: function (statusNumber: number) {
      expect(statusNumber).toBe(201)
      return this
    },
    send: () => this,
  }
  await signup(req, res)
})

afterAll(async () => {
  await prisma.$executeRaw(
    `TRUNCATE TABLE public."Comment" RESTART IDENTITY CASCADE;`,
  )
  await prisma.$executeRaw(
    `TRUNCATE TABLE public."Post" RESTART IDENTITY CASCADE;`,
  )
  await prisma.$executeRaw(
    `TRUNCATE TABLE public."User" RESTART IDENTITY CASCADE;`,
  )
  await prisma.$disconnect()
})

describe('test post controller happy way', () => {
  test('create post is valid', async () => {
    await new Promise((r) => setTimeout(r, 2000))

    const req: any = reqGlobal

    const res: any = {
      status: function (statusNumber: number) {
        expect(statusNumber).toBe(200)
        return this
      },
      json: function (jsonResult: any) {
        expect(jsonResult.post.title).toBe('title of fake post')
        return this
      },
    }
    await createPost(req, res)
  })

  test('find post is valid', async () => {
    const req: any = reqGlobal

    const res: any = {
      status: function (statusNumber: number) {
        expect(statusNumber).toBe(200)
        return this
      },
      json: function (jsonResult: any) {
        expect(jsonResult.posts[0].title).toBe('title of fake post')
        return this
      },
    }
    await findPosts(req, res)
  })

  test('find specific post is valid', async () => {
    const req: any = reqGlobal

    const res: any = {
      status: function (statusNumber: number) {
        expect(statusNumber).toBe(200)
        return this
      },
      json: function (jsonResult: any) {
        expect(jsonResult.post.title).toBe('title of fake post')
        return this
      },
    }
    await findPost(req, res)
  })

  test('update specific post is valid', async () => {
    const req: any = {
      ...reqGlobal,
      body: { ...reqGlobal.body, title: 'updated title' },
    }

    const res: any = {
      status: function (statusNumber: number) {
        expect(statusNumber).toBe(200)
        return this
      },
      json: function (jsonResult: any) {
        expect(jsonResult.post.title).toBe('updated title')
        return this
      },
    }
    await updatePost(req, res)
  })

  test('publish specific post is valid', async () => {
    const req: any = reqGlobal

    const res: any = {
      status: function (statusNumber: number) {
        expect(statusNumber).toBe(200)
        return this
      },
      json: function (jsonResult: any) {
        expect(jsonResult.post.published).toBe(true)
        return this
      },
    }
    await publishPost(req, res)
  })

  test('like specific post is valid', async () => {
    const req: any = {
      ...reqGlobal,
      body: { ...reqGlobal.body, liked: [] },
    }

    const res: any = {
      status: function (statusNumber: number) {
        expect(statusNumber).toBe(200)
        return this
      },
      json: function (jsonResult: any) {
        expect(jsonResult.success).toBe(true)
        return this
      },
    }
    await likedPost(req, res)
  })

  test('unlike specific post is valid', async () => {
    const req: any = {
      ...reqGlobal,
      body: { ...reqGlobal.body, liked: [1] },
    }

    const res: any = {
      status: function (statusNumber: number) {
        expect(statusNumber).toBe(200)
        return this
      },
      json: function (jsonResult: any) {
        expect(jsonResult.success).toBe(true)
        return this
      },
    }
    await likedPost(req, res)
  })

  test('delete specific post is valid', async () => {
    const req: any = reqGlobal

    const res: any = {
      status: function (statusNumber: number) {
        expect(statusNumber).toBe(200)
        return this
      },
      json: function (jsonResult: any) {
        expect(jsonResult.post.title).toBe('updated title')
        return this
      },
    }
    await deletePost(req, res)
  })
})

describe('test post controller error way', () => {
  test('create post with empty data is invalid', async () => {
    const req: any = {}

    const res: any = {
      status: function (statusNumber: number) {
        expect(statusNumber).toBe(404)
        return this
      },
      json: function (jsonResult: any) {
        expect(jsonResult.error).toMatchSnapshot()
        return this
      },
      end: () => this,
    }
    await createPost(req, res)
  })

  test('find empty posts is invalid', async () => {
    const req: any = {}

    const res: any = {
      status: function (statusNumber: number) {
        expect(statusNumber).toBe(404)
        return this
      },
      json: function (jsonResult: any) {
        expect(jsonResult.error).toMatchSnapshot()
        return this
      },
      end: () => this,
    }
    await findPosts(req, res)
  })

  test('find empty specific post is invalid', async () => {
    const req: any = { ...reqGlobal, params: { id: 100 } }

    const res: any = {
      status: function (statusNumber: number) {
        expect(statusNumber).toBe(404)
        return this
      },
      json: function (jsonResult: any) {
        expect(jsonResult.error).toMatchSnapshot()
        return this
      },
      end: () => this,
    }
    await findPost(req, res)
  })

  test('update empty specific post is invalid', async () => {
    const req: any = {
      ...reqGlobal,
      params: { id: 100 },
      body: { ...reqGlobal.body, title: 'updated title' },
    }

    const res: any = {
      status: function (statusNumber: number) {
        expect(statusNumber).toBe(403)
        return this
      },
      json: function (jsonResult: any) {
        expect(jsonResult.error).toMatchSnapshot()
        return this
      },
      end: () => this,
    }
    await updatePost(req, res)
  })

  test('publish empty specific post is isvalid', async () => {
    const req: any = { ...reqGlobal, params: { id: 100 } }

    const res: any = {
      status: function (statusNumber: number) {
        expect(statusNumber).toBe(403)
        return this
      },
      json: function (jsonResult: any) {
        expect(jsonResult.error).toMatchSnapshot()
        return this
      },
      end: () => this,
    }
    await publishPost(req, res)
  })

  test('like empty specific post is invalid', async () => {
    const req: any = { ...reqGlobal, params: { id: 100 } }

    const res: any = {
      status: function (statusNumber: number) {
        expect(statusNumber).toBe(404)
        return this
      },
      json: function (jsonResult: any) {
        expect(jsonResult.error).toMatchSnapshot()
        return this
      },
      end: () => this,
    }
    await likedPost(req, res)
  })

  test('delete empty specific post is invalid', async () => {
    const req: any = { ...reqGlobal, params: { id: 100 } }

    const res: any = {
      status: function (statusNumber: number) {
        expect(statusNumber).toBe(403)
        return this
      },
      json: function (jsonResult: any) {
        expect(jsonResult.error).toMatchSnapshot()
        return this
      },
      end: () => this,
    }
    await deletePost(req, res)
  })
})
