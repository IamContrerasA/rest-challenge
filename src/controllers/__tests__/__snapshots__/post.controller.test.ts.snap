// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test post controller error way create post with empty data is invalid 1`] = `"Couldn't create a post"`;

exports[`test post controller error way delete empty specific post is invalid 1`] = `"Current user cannot edit the information of another user, only that of him"`;

exports[`test post controller error way find empty posts is invalid 1`] = `"Couldn't find any posts jet"`;

exports[`test post controller error way find empty specific post is invalid 1`] = `"Couldn't find pos with id = '100'"`;

exports[`test post controller error way like empty specific post is invalid 1`] = `"Couldn't find pos with id = '100'"`;

exports[`test post controller error way publish empty specific post is isvalid 1`] = `"Current user cannot edit the information of another user, only that of him"`;

exports[`test post controller error way update empty specific post is invalid 1`] = `"Current user cannot edit the information of another user, only that of him"`;
